sys/net/crosslayer/ng_pktbuf/ng_pktbuf.c

```diff
@@ -87,16 +103,21 @@ void ng_pktbuf_release(ng_pktsnip_t *pkt)
         return;
     }
 
-    atomic_set_return(&(pkt->users), pkt->users - 1);
+    mutex_lock(&_pktbuf_mutex);
 
-    if (pkt->users == 0 && _pktbuf_internal_contains(pkt->data)) {
-        mutex_lock(&_pktbuf_mutex);
+    while (pkt != NULL) {
+        pkt->users--;
 
-        _pktbuf_internal_free(pkt->data);
-        _pktbuf_internal_free(pkt);
+        if (pkt->users == 0 && _pktbuf_internal_contains(pkt->data)) {
```

Comment:
don't we need to call `_pktbuf_internal_free(pkt)` always if `pkt->users == 0`, independent of `pkt->data`? So this if condition needs to be split up in my opinion:
```c
if (pkt->users == 0) {
    _pktbuf_internal_free(pkt);
    if (_pktbuf_internal_contains(pkt->data)) {
        _pktbuf_internal_free(pkt);
    }
}
```